lib/spack/spack/spec.py

```diff
@@ -3451,25 +3491,123 @@ def _autospec(self, spec_like):
             return spec_like
         return Spec(spec_like)
 
-    def satisfies(self, other, deps=True, strict=False):
-        """Determine if this spec satisfies all constraints of another.
-
-        There are two senses for satisfies, depending on the ``strict``
-        argument.
+    def intersects(self, other: "Spec", deps: bool = True) -> bool:
+        """Return True if there exists at least one concrete spec that matches both
+        self and other, otherwise False.
 
-          * ``strict=False``: the left-hand side and right-hand side have
-            non-empty intersection. For example ``zlib`` satisfies
-            ``zlib@1.2.3`` and ``zlib@1.2.3`` satisfies ``zlib``. In this
-            sense satisfies is a commutative operation: ``x.satisfies(y)``
-            if and only if ``y.satisfies(x)``.
+        This operation is commutative, and if two specs intersect it means that one
+        can constrain the other.
 
-          * ``strict=True``: the left-hand side is a subset of the right-hand
-            side. For example ``zlib@1.2.3`` satisfies ``zlib``, but ``zlib``
-            does not satisfy ``zlib@1.2.3``. In this sense satisfies is not
-            commutative: the left-hand side should be at least as constrained
-            as the right-hand side.
+        Args:
+            other: spec to be checked for compatibility
+            deps: if True check compatibility of dependency nodes too, if False only check root
         """
+        other = self._autospec(other)
+
+        # Optimizations concrete specs
+        if other.concrete and self.concrete:
+            return self.dag_hash() == other.dag_hash()
+
+        # TODO (INTERSECTS): the lines below are not commutative
```

Comment:
This is one place where I need to look what needs to be done. This is the code we had before, but it is clearly non-commutative.

Reply:
Fixed in aea2da1aea8f3245f95fc40335a740ff66acfa06